### PR TITLE
several bug fixes

### LINKED ISSUE #1546: Farmer doesn´t do anything
### Minecolonies version

version 0.8.4369 minecraft 1.12.2

### Expected behavior

- Farmer take crops, and replant.

### Actual behaviour

- Anything, it stay´s stand.

### Steps to reproduce the problem

1. Build the Farm 
2. Asign the farmer
3. I need to uste the Recall Worker button several times to make him plant crops, but then it stays when the farm is all planted and doesn´t pick them when are already mature.
4. It stays forever, and I tried to refresh the colony (with the "/mc colony refresh") and even killed the farmer and nothing.

### LINKED ISSUE #1579: Enchantments bug.
### Expected behavior

- give colonist enchanted weapon they cant use and they dont use it.

### Actual behaviour

- uses weapon they cant use.

### Steps to reproduce the problem

1. Heavily enchant weapon 1 and give to colonist. They still complain they need a weapon.
2. Give colonist an unenchanted version of same weapon but leave enchanted one in inventory to the LEFT of the unenchanted one.
3. Colonist uses the enchanted weapon.
4. Profit

My guard was asking for a diamond sword. Gave him one with every enchant on it and he still asked for a diamond sword. Gave him a plain diamond sword and he ran off setting everything he attacked on fire and the enchanted sword took durability loss.

Same thing happened with a ranger. Wanted a bow with max enchant level 3. Same result as previous example.

### LINKED ISSUE #1583: Death of Citizen - Suggestion
Would be nice in the chat log to indicate how the citizen died,  just instead of coord and they are dead.  Would like to know what got it.

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #1579
Closes #1546
Closes #1583

# Changes proposed in this pull request:
- Fixes problems where guards use OP weapons for their level
- Improve randomness of the farmer
- Fix gathering of guards
- Fix problems with housing of barracks
- Explain citizend eath
Review please

### CODE REVIEW

**Review by marvin-bitterlich — APPROVED** [review]
nice changes! I like getting rid of the instanceof

**Review by marchermans — APPROVED** [review]

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 63 issues
* ![MAJOR][MAJOR] 23 major
* ![MINOR][MINOR] 38 minor
* ![INFO][INFO] 2 info

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L121](https://github.com/Minecolonies/minecolonies/blob/09b1e13c753dd2ce82b4f8ae8ceaab5909eb8fd6/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L121): Remove this unused method parameter "stack". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1172)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L192](https://github.com/Minecolonies/minecolonies/blob/09b1e13c753dd2ce82b4f8ae8ceaab5909eb8fd6/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L192): Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L343](https://github.com/Minecolonies/minecolonies/blob/09b1e13c753dd2ce82b4f8ae8ceaab5909eb8fd6/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L343): Reduce this class from 52 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [BuildingFarmer.java#L552](https://github.com/Minecolonies/minecolonies/blob/09b1e13c753dd2ce82b4f8ae8ceaab5909eb8fd6/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingFarmer.java#L552): Reduce this class from 86 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [EntityCitizen.java](https://github.com/Minecolonies/minecolonies/blob/09b1e13c753dd2ce82b4f8ae8ceaab5909eb8fd6/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java): This file has 1,424 lines, which is greater than 1,000 authorized. Split it into smaller files. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L72](https://github.com/Minecolonies/minecolonies/blob/09b1e13c753dd2ce82b4f8ae8ceaab5909eb8fd6/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L72): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 61 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L72](https://github.com/Minecolonies/minecolonies/blob/09b1e13c753dd2ce82b4f8ae8ceaab5909eb8fd6/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L72): class "EntityCitizen" has 100 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L382](https://github.com/Minecolonies/minecolonies/blob/09b1e13c753dd2ce82b4f8ae8ceaab5909eb8fd6/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L382): Assign this magic number 3.0F to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L383](https://github.com/Minecolonies/minecolonies/blob/09b1e13c753dd2ce82b4f8ae8ceaab5909eb8fd6/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L383): Assign this magic number 5.0F to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L383](https://github.com/Minecolonies/minecolonies/blob/09b1e13c753dd2ce82b4f8ae8ceaab5909eb8fd6/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L383): Assign this magic number 0.02F to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 63 issues
* ![MAJOR][MAJOR] 23 major
* ![MINOR][MINOR] 38 minor
* ![INFO][INFO] 2 info

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L121](https://github.com/Minecolonies/minecolonies/blob/09b1e13c753dd2ce82b4f8ae8ceaab5909eb8fd6/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L121): Remove this unused method parameter "stack". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1172)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L192](https://github.com/Minecolonies/minecolonies/blob/09b1e13c753dd2ce82b4f8ae8ceaab5909eb8fd6/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L192): Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L343](https://github.com/Minecolonies/minecolonies/blob/09b1e13c753dd2ce82b4f8ae8ceaab5909eb8fd6/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L343): Reduce this class from 52 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [BuildingFarmer.java#L552](https://github.com/Minecolonies/minecolonies/blob/09b1e13c753dd2ce82b4f8ae8ceaab5909eb8fd6/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingFarmer.java#L552): Reduce this class from 86 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [EntityCitizen.java](https://github.com/Minecolonies/minecolonies/blob/09b1e13c753dd2ce82b4f8ae8ceaab5909eb8fd6/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java): This file has 1.424 lines, which is greater than 1.000 authorized. Split it into smaller files. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L72](https://github.com/Minecolonies/minecolonies/blob/09b1e13c753dd2ce82b4f8ae8ceaab5909eb8fd6/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L72): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 61 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L72](https://github.com/Minecolonies/minecolonies/blob/09b1e13c753dd2ce82b4f8ae8ceaab5909eb8fd6/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L72): class "EntityCitizen" has 100 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L382](https://github.com/Minecolonies/minecolonies/blob/09b1e13c753dd2ce82b4f8ae8ceaab5909eb8fd6/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L382): Assign this magic number 3.0F to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L383](https://github.com/Minecolonies/minecolonies/blob/09b1e13c753dd2ce82b4f8ae8ceaab5909eb8fd6/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L383): Assign this magic number 5.0F to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L383](https://github.com/Minecolonies/minecolonies/blob/09b1e13c753dd2ce82b4f8ae8ceaab5909eb8fd6/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L383): Assign this magic number 0.02F to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java [hunk 1]
```diff
@@ -947,7 +947,7 @@ public boolean hasPermission(final UUID id, @NotNull final Action action)
         public boolean hasPermission(final Rank rank, @NotNull final Action action)
         {
             return (rank == Rank.OWNER && action != Action.GUARDS_ATTACK)
-                     || (permissions != null && action != null && Utils.testFlag(permissions.get(rank), action.getFlag()));
+                     || (permissions != null && action != null && permissions.containsKey(rank) && Utils.testFlag(permissions.get(rank), action.getFlag()));
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java:950]: MAJOR Remove this expression which always evaluates to "true" [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2589)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 68 issues
* ![MAJOR][MAJOR] 28 major
* ![MINOR][MINOR] 38 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L121](https://github.com/Minecolonies/minecolonies/blob/57dab722cf9f8eace777ec321d7b62b70a55f188/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L121): Remove this unused method parameter "stack". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1172)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L192](https://github.com/Minecolonies/minecolonies/blob/57dab722cf9f8eace777ec321d7b62b70a55f188/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L192): Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [AbstractBuildingWorker.java#L343](https://github.com/Minecolonies/minecolonies/blob/57dab722cf9f8eace777ec321d7b62b70a55f188/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L343): Reduce this class from 52 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [BuildingFarmer.java#L552](https://github.com/Minecolonies/minecolonies/blob/57dab722cf9f8eace777ec321d7b62b70a55f188/src/main/java/com/minecolonies/coremod/colony/buildings/BuildingFarmer.java#L552): Reduce this class from 86 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [Permissions.java#L29](https://github.com/Minecolonies/minecolonies/blob/57dab722cf9f8eace777ec321d7b62b70a55f188/src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java#L29): class "Permissions" has 36 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [Permissions.java#L325](https://github.com/Minecolonies/minecolonies/blob/57dab722cf9f8eace777ec321d7b62b70a55f188/src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java#L325): Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [Permissions.java#L786](https://github.com/Minecolonies/minecolonies/blob/57dab722cf9f8eace777ec321d7b62b70a55f188/src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java#L786): Remove this unused private "isSubscriber" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUnusedPrivateMethod)
1. ![MAJOR][MAJOR] [Permissions.java#L871](https://github.com/Minecolonies/minecolonies/blob/57dab722cf9f8eace777ec321d7b62b70a55f188/src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java#L871): Reduce this class from 116 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)
1. ![MAJOR][MAJOR] [EntityCitizen.java](https://github.com/Minecolonies/minecolonies/blob/57dab722cf9f8eace777ec321d7b62b70a55f188/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java): This file has 1,424 lines, which is greater than 1,000 authorized. Split it into smaller files. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [EntityCitizen.java#L72](https://github.com/Minecolonies/minecolonies/blob/57dab722cf9f8eace777ec321d7b62b70a55f188/src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java#L72): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 61 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by Moobien** [comment]
I wish there was more i could do to help you guys other than finding more work for you to do

**Pull request merged**